File: rob599_hw3/scripts/timer.py
#!/usr/bin/env python

import sys
import rospy
import math
import time
import numpy as np
from sensor_msgs.msg import LaserScan
#######################################
# Laser Scan: 
#   Header: Seq, Stamp, frame_id
#   Angle_min, Angle_max, Angle_Increment, Time_Increment
#   Scan time, range_min, range_max, ranges, intensities 
#######################################

class Timer:
    def __init__(self, argv):
        self.laser_sub = rospy.Subscriber('/base_scan', LaserScan, self.laser_callback)
        self.laser_sub = rospy.Subscriber('/node_time_taken', LaserScan, self.time_callback)
        self.scan_pub = rospy.Publisher('/timer_start', LaserScan, queue_size= 1)
        self.counter = 0
        self.print_ct = 0
        self.average_time = 0
        self.complete = 0
        self.filepath = argv[0]
        if self.filepath.find('scripts/timer.py') != -1:
            self.filepath = self.filepath.replace('scripts/timer.py', 'config/')
        elif self.filepath.find('config') != -1:
            self.filepath = self.filepath.replace('/rob599_hw3', '/rob599_hw3/config/')
            self.filepath = self.filepath.replace('//', '/')
        if rospy.get_param('node_num') == 1:
            self.node = "Timer node to Timer node"
        elif rospy.get_param('node_num') == 2:
            self.node = "Timer node to Median node to Timer node"
        elif rospy.get_param('node_num') == 3:
            self.node = "Timer node to Median node to Noiser node to Timer node"


    def time_callback(self, msg):
        self.average_time += time.time() - self.counter
        self.complete = 0
        print('Python: Time taken by {0} is {1} secs'.format(self.node, time.time() - self.counter))
        self.print_ct  += 1
        if self.print_ct == 100:
            print('Python: Time taken by {0} for 100 times is {1} secs'.format(self.node, self.average_time/100))
            with open(self.filepath+'time.txt', 'a') as self.csv_writer:
                self.csv_writer.write('\nPython: Time taken by {0} for 100 times is {1} secs\n'.format(self.node, self.average_time/100))


    def laser_callback(self, msg):
        self.scan_pub.publish(msg)
        if self.complete == 0:
            self.counter = time.time()
            self.complete = 1


if __name__ == '__main__':
    rospy.init_node('timer', anonymous=True)
    if sys.argv != 0:
        timer = Timer(sys.argv)
    else:
        rospy.logerr('System arguments are not given')
    try:
        rospy.spin()
    except KeyboardInterrupt:
        print("Shutting down")

File: rob599_hw3/src/timer.cpp
#include "ros/ros.h"
#include "std_msgs/String.h"
#include "sensor_msgs/LaserScan.h"
#include <iostream>
#include <iterator>
#include <random>
#include <sstream>
#include <fstream>
#include <string>
#include <ros/package.h>

class Timer 
{
  public:
    double secs;
    int complete;
    double average_time;
    std::string filepath;
    int print = 0;
    ros::NodeHandle n;
    ros::Publisher timer_pub = n.advertise<sensor_msgs::LaserScan>("/timer_start", 1000);
    void laser_Callback(const sensor_msgs::LaserScan::ConstPtr& msg);
    void time_Callback(const sensor_msgs::LaserScan::ConstPtr& msg);
};


void Timer::laser_Callback(const sensor_msgs::LaserScan::ConstPtr& msg)
{
  sensor_msgs::LaserScan filtered_scan;
  filtered_scan.header = msg->header;
  filtered_scan.angle_max = msg->angle_max;
  filtered_scan.angle_min = msg->angle_min;
  filtered_scan.ranges = msg->ranges;
  filtered_scan.angle_increment = msg->angle_increment;
  filtered_scan.time_increment = msg->time_increment;
  filtered_scan.scan_time = msg->scan_time;
  filtered_scan.intensities = msg->intensities;
  filtered_scan.range_min = msg->range_min;
  filtered_scan.range_max = msg->range_max;
  filtered_scan.ranges = msg -> ranges;
  timer_pub.publish(filtered_scan);
  if (complete ==0){
  complete = 1;
  secs =ros::Time::now().toSec();
  }
  }

void Timer::time_Callback(const sensor_msgs::LaserScan::ConstPtr& msg)
{
  complete = 0;
  average_time = average_time + (ros::Time::now().toSec() - secs);
  print = print + 1;
  ROS_INFO("C++: Time taken by nodes is %f", (ros::Time::now().toSec() - secs));
  if(print == 100)
  {
    ROS_INFO("C++: Time taken by nodes for 100 steps is %f", average_time/100);
    std::ofstream outfile;
    outfile.open(filepath, std::ios_base::app);
    outfile << "\nC++: Time taken by nodes for 100 ste is " << (average_time/100) << "\n";
  }
  }

int main(int argc, char **argv)
{
  ros::init(argc, argv, "timer");
  ros::NodeHandle n;
  std::string str1 ("rob599_hw3");
  std::string str2 ("rob599_hw3/config/time.txt");
  std::string path = ros::package::getPath("rob599_hw3");
  Timer timer;
  path.replace(path.find(str1), str1.length(), str2);
  timer.filepath = path;
  timer.complete = 0;
  timer.average_time = 0;
  ros::Subscriber sub = n.subscribe<sensor_msgs::LaserScan>("/base_scan", 1000, &Timer::laser_Callback, &timer);
  ros::Subscriber sub1 = n.subscribe<sensor_msgs::LaserScan>("/node_time_taken", 1000, &Timer::time_Callback, &timer);
  ros::spin();
  return 0;
}
